### PR TITLE
Local time stepping (time steppers)

### PR DESCRIPTION
**Pull request by wthrowe**
This adds support to the AdamsBashforthN TimeStepper for local time stepping.  RungeKutta3 continues to only work with global time stepping, but implements the same interface.  This does not cover the driver and communication routines because we don't have the infrastructure for those in place yet.

### CODE REVIEW

**Comment by wthrowe** [comment]
Known question: Where does LagrangePolynomial belong?

File: src/DataStructures/GeneralIndexIterator.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+#pragma once
+
+#include <type_traits>
+#include <vector>
+
+/// Iterate over a generic "block" of coordinates.  The constructor
+/// takes a container of pairs defining the range (inclusive start,
+/// exclusive end) of each coordinate.  The utility function
+/// make_GeneralIndexIterator is provided for easier construction.
```
Comment (kidder) [inline src/DataStructures/GeneralIndexIterator.hpp]: check doc (may need ())

File: src/DataStructures/GeneralIndexIterator.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      ++current_[index];
+      if (current_[index] != ranges_[index].second) { return; }
+      current_[index] = ranges_[index].first;
+    }
+    done_ = true;
+  }
+
+  /// Iterate over the coordinates of the current value.
+  //@{
+  iterator_type cbegin() const { return current_.cbegin(); }
```
Comment (kidder) [inline src/DataStructures/GeneralIndexIterator.hpp]: noexcept everything that is noexcept

File: src/DataStructures/GeneralIndexIterator.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+ private:
+  Container ranges_;
+  std::vector<value_type> current_{};
+  bool done_{false};
+};
+
+/// Construct a GeneralIndexIterator
+template <typename Container>
+GeneralIndexIterator<std::decay_t<Container>> make_GeneralIndexIterator(
```
Comment (kidder) [inline src/DataStructures/GeneralIndexIterator.hpp]: make_general_...

File: src/Numerical/Spectral/LagrangePolynomial.hpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+#pragma once
+
+#include <iterator>
+
+#include "ErrorHandling/Assert.hpp"
+
+/// Evaluate the jth Lagrange interpolating polynomial with the given
+/// control points where j is the index of index_point.
+template <typename Iterator>
+double LagrangePolynomial(const Iterator& index_point,
```
Comment (kidder) [inline src/Numerical/Spectral/LagrangePolynomial.hpp]: lagrange_polynomial

File: src/Numerical/Spectral/LagrangePolynomial.hpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    const double x_m = *m_it;
+    result *= (x_m - x) / (x_m - x_j);
+  }
+  return result;
+}
+
+/// Evaluate the jth (zero-indexed) Lagrange interpolating polynomial
+/// with the given control points.
+template <typename Iterator>
+double LagrangePolynomial(size_t j, double x,
```
Comment (kidder) [inline src/Numerical/Spectral/LagrangePolynomial.hpp]: snake_case

File: tests/Unit/Numerical/Spectral/Test_LagrangePolynomial.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+SPECTRE_TEST_CASE("Unit.Numerical.Spectral.LagrangePolynomial",
+                  "[Unit][Numerical]") {
+  std::array<double, 4> control{{-1., 0., 1.5, 6.}};
+
+  for (size_t i = 0; i < control.size(); ++i) {
+    for (size_t j = 0; j < control.size(); ++j) {
+      CHECK(LagrangePolynomial(j, control[i], control.begin(), control.end()) ==
+            approx(i == j ? 1. : 0.));
+      CHECK(LagrangePolynomial(std::next(control.begin(),
```
Comment (kidder) [inline tests/Unit/Numerical/Spectral/Test_LagrangePolynomial.cpp]: use + instead of next

File: src/Utilities/CachedFunction.hpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+/// A function wrapper that caches function values.
+template <typename Function, typename Map>
+class CachedFunction {
+ public:
+  using Input = typename Map::key_type;
```
Comment (kidder) [inline src/Utilities/CachedFunction.hpp]: input

File: src/Utilities/CachedFunction.hpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  using Output = typename Map::mapped_type;
```
Comment (kidder) [inline src/Utilities/CachedFunction.hpp]: output

File: src/Utilities/CachedFunction.hpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+///
+/// \example
+/// \snippet Test_CachedFunction.cpp make_CachedFunction_example
+///
+/// \tparam Input function argument type
+/// \tparam Map class template to use as the map holding the cache
+/// \param map_args arguments to pass to the map constructor
+template <typename Input, template <typename...> class Map = std::unordered_map,
+          typename... MapArgs, typename Function, typename... PassedMapArgs>
+auto make_CachedFunction(Function function,
```
Comment (kidder) [inline src/Utilities/CachedFunction.hpp]: make_cached_function

File: src/Utilities/CachedFunction.hpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                         PassedMapArgs... map_args) noexcept {
+  using Output = std::result_of_t<Function&(const Input&)>;
```
Comment (kidder) [inline src/Utilities/CachedFunction.hpp]: output

File: tests/Unit/Utilities/Test_CachedFunction.cpp [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  };
+
+  bool comparator_used = false;
+  const auto comparator = [&comparator_used](const std::string& a,
+                                             const std::string& b) noexcept {
+    comparator_used = true;
+    return a < b;
+  };
+
+  // [make_CachedFunction_example]
```
Comment (kidder) [inline tests/Unit/Utilities/Test_CachedFunction.cpp]: check if need ///

**Review by kidder — CHANGES_REQUESTED** [review]
for some of the commits

**Comment by wthrowe** [comment]
Addressed comments and clang-tidy, reset all the author dates so the commits show up in order in GitHub.  Changes to commits that have not been reviewed yet were immediately squashed.

One of the Travis builds had (and presumably will again have) a -Werror failure related to boost::optional that appears to just be the compiler's static analyzer not understanding the class.  I'm not sure of a way to fix that other than adding warning pragmas.

File: src/Time/Time.hpp [hunk 12]
```diff
@@ -57,6 +57,25 @@ class Time {
   // clang-tidy: google-runtime-references
   void pup(PUP::er& p) noexcept;  // NOLINT
 
+  /// A comparison operator that compares Times structurally, i.e.,
+  /// just looking at the class members.  The returned ordering does
+  /// not match the time ordering and opposite sides of slab
+  /// boundaries do not compare equal.  It is, however, much faster to
+  /// compute than the temporal ordering, so is useful when an
+  /// ordering is required, but the ordering does not have to be
+  /// physically meaningful.
```
Comment (nilsdeppe) [inline src/Time/Time.hpp]: "So this can be used as the key to a map"

File: src/Time/TimeSteppers/TimeStepper.hpp [hunk 13]
```diff
@@ -59,6 +62,35 @@ class TimeStepper : public Factory<TimeStepper> {
...
+  /// The first element of `history` should represent the local side.
+  /// The contents of each element should be ordered in the same
+  /// manner as for `update_u`.
+  ///
+  /// The coupling function `coupling` should compute the derivative
+  /// of the boundary quantity from the side data, supplied as a
+  /// `std::vector<std::reference_wrapper<const FluxVars>>`.  These
+  /// values may be used to form a linear combination internally, so
+  /// `BoundaryVars` should have appropriate mathematical operators
+  /// defined to allow that.
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/TimeStepper.hpp:84]: `vector` -> neighbors, first value is self

`deque` -> time history

File: src/Time/TimeSteppers/RungeKutta3.hpp [hunk 14]
```diff
@@ -43,23 +48,64 @@ class RungeKutta3 : public TimeStepper::Inherit {
...
+  return step_work(u, history, time_step, std::get<2>(history.back()));
+}
+
+template <typename BoundaryVars, typename FluxVars, typename Coupling>
+BoundaryVars RungeKutta3::compute_boundary_delta(
+    const Coupling& coupling,
+    const std::vector<std::deque<std::tuple<Time, BoundaryVars, FluxVars>>>&
+        history,
+    const TimeDelta& time_step) const noexcept {
+  std::vector<std::reference_wrapper<const FluxVars>> coupling_args;;
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/RungeKutta3.hpp]: remove extra `;`

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 15]
```diff
@@ -98,6 +101,44 @@ class AdamsBashforthN : public TimeStepper::Inherit {
...
+
+    bool operator()(const Time& a, const Time& b) const noexcept {
+      return forward_in_time_ ? a < b : b < a;
+    }
+
+   private:
+    bool forward_in_time_;
+  };
+
+  /// Convert an iterator over history entries to one over their times.
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: one -> an iterator

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 16]
```diff
@@ -98,6 +101,44 @@ class AdamsBashforthN : public TimeStepper::Inherit {
...
+  static auto time_iter(const Iterator& it) noexcept {
+    // Cast the overloaded get function to a function pointer taking a
+    // history entry to a time reference.  This is needed to prevent the
+    // boost::make_transform_iterator calls from being ambiguous.
+    const Time& (*const get_time)(const typename Iterator::value_type&) =
+        std::get<0>;
+    return boost::make_transform_iterator(it, get_time);
+  }
+
+  /// Find the step on a give side containing the given time.  Steps
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: give -> given

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 17]
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
...
+  for (const auto& sh : history) {
+    ASSERT(not sh.empty(), "No data");
+    ASSERT(std::is_sorted(time_iter(sh.begin()), time_iter(sh.end()),
+                          simulation_less),
+           "History not in order");
+  }
+
+  const size_t num_sides = history.size();
+
+  // Start and end of the stepwe are trying to take
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: step we

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 18]
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
...
+  auto ab_coefs = [&all_times, target_order_s]() noexcept {
+    using Iter = decltype(all_times.cbegin());
+    auto compare = [](const Iter& a, const Iter& b) noexcept {
+      return Time::StructuralCompare{}(*a, *b);
+    };
+    return make_cached_function<Iter, std::map, decltype(compare)>(
+        [target_order = target_order_s](const Iter& it) noexcept {
+          const auto it_end = std::next(it);
+          const auto it_start = std::prev(it, target_order - 1);
+          return TimeSteppers::AdamsBashforthN::get_coefficients(
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: reduce qualifiers?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Time/Time.hpp [hunk 19]
```diff
@@ -58,12 +58,13 @@ class Time {
...
 
   /// A comparison operator that compares Times structurally, i.e.,
-  /// just looking at the class members.  The returned ordering does
-  /// not match the time ordering and opposite sides of slab
-  /// boundaries do not compare equal.  It is, however, much faster to
-  /// compute than the temporal ordering, so is useful when an
-  /// ordering is required, but the ordering does not have to be
-  /// physically meaningful.
+  /// just looking at the class members.  This is only intended for
+  /// use as the comparitor in a map.  The returned ordering does not
```
Comment (fmahebert) [inline src/Time/Time.hpp]: comparator

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 20]
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
...
+           make_general_index_iterator(std::move(coupling_evaluation_ranges));
+       coupling_evaluation;
+       ++coupling_evaluation) {
+    double deriv_coef = 0.;
+
+    // Prepare to loop over smallest-scale segments relevant for the
+    // current coupling evaluation.
+
+    // Latest of start of current step (otherwise computed in a
+    // preious call) and the times of the side data being coupled
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: previous call

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 21]
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
...
+      }
+    }
+    const auto smallest_scale_times_begin =
+        std::lower_bound(all_times.cbegin(), all_times.cend(),
+                         coupling_start_time, simulation_less);
+    const auto smallest_scale_times_end =
+        std::lower_bound(smallest_scale_times_begin, all_times.cend(),
+                         coupling_end_time, simulation_less);
+
+    // Step on each side containing the boundary step we are currently
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: "The step on each side"

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 22]
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
...
+    for (const auto& side_hist : history) {
+      side_steps.push_back(step_containing_time(side_hist,
+                                                *smallest_scale_times_begin,
+                                                simulation_less));
+    }
+
+    for (auto smallest_scale_time = smallest_scale_times_begin;
+         smallest_scale_time != smallest_scale_times_end;
+         ++smallest_scale_time) {
+      // Advance any side step where we've moved into the next one.
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: Advance the step of any side where we've moved into the next step

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 23]
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
...
+
+        // Multiply ab_term by coefficients to interpolate each side
+        // to evaluation time.
+        for (size_t side = 0; side < num_sides; ++side) {
+          // Stop calculating things if we're just going to multiply
+          // them by zero.
+          if (ab_term == 0) { break; }
+
+          // Makes an iterator with a map to give time as a double.
+          const auto make_lagrange_iterator = [&](const HistoryIterator& it) {
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: only capture what's needed

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 24]
```diff
@@ -188,6 +225,217 @@ TimeDelta AdamsBashforthN::update_u(
...
+      }
+
+      if (accumulated_change) {
+        *accumulated_change += deriv_coef * coupling(args);
+      } else {
+        accumulated_change =
+            boost::optional<BoundaryVars>(deriv_coef * coupling(args));
+      }
+    }
+  }
```
Comment (nilsdeppe) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: add namespace style comments at end of long for loops

File: tests/Unit/Time/TimeSteppers/Test_AdamsBashforthN.cpp [hunk 25]
```diff
@@ -74,3 +79,281 @@ SPECTRE_TEST_CASE("Unit.Time.TimeSteppers.AdamsBashforthN.Factory",
...
+NCd& operator+=(NCd& a, NCd&& b) { return a = NCd(a() + b()); }
+
+// Random numbers
+constexpr double c10 = 0.949716728952811;
+constexpr double c11 = 0.190663110072823;
+constexpr double c20 = 0.932407227651314;
+constexpr double c21 = 0.805454101952822;
+constexpr double c22 = 0.825876851406978;
+
+double quartic_side1(double x) { return c10 + x * c11; }
```
Comment (nilsdeppe) [inline tests/Unit/Time/TimeSteppers/Test_AdamsBashforthN.cpp]: how are these related to each other?

File: tests/Unit/Time/TimeSteppers/Test_AdamsBashforthN.cpp [hunk 26]
```diff
@@ -74,3 +79,281 @@ SPECTRE_TEST_CASE("Unit.Time.TimeSteppers.AdamsBashforthN.Factory",
...
+        history[side].emplace_front(
+            now, NCd(unused), NCd(gsl::at(side_solutions, side)(now.value())));
+      }
+    }
+  }
+
+  double y = 0.;
+  while (t < slab.end()) {
+    for (size_t side = 0; side < 3; ++side) {
+    history[side].emplace_back(t, NCd(unused),
```
Comment (nilsdeppe) [inline tests/Unit/Time/TimeSteppers/Test_AdamsBashforthN.cpp]: indent

File: tests/Unit/Time/TimeSteppers/TimeStepperTestUtils.hpp [hunk 27]
```diff
@@ -97,6 +95,8 @@ void integrate_test(const Stepper& stepper, const double integration_time,
     // This check needs a looser tolerance for lower-order time steppers.
     CHECK(y == approx(analytic(time.value())).epsilon(epsilon));
   }
+  // Make sure history is being cleaned up.
```
Comment (nilsdeppe) [inline tests/Unit/Time/TimeSteppers/TimeStepperTestUtils.hpp]: Mention that twenty is arbitrary, won't be exceeded in general cases

File: src/Time/Time.hpp [hunk 28]
```diff
@@ -58,12 +58,13 @@ class Time {
...
+  /// match the time ordering and opposite sides of slab boundaries do
+  /// not compare equal.  It is, however, much faster to compute than
+  /// the temporal ordering, so is useful when an ordering is
```
Comment (nilsdeppe) [inline src/Time/Time.hpp]: so it's useful

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Comment by wthrowe** [comment]
Trying to silence the warnings causing the Travis build failure with pragmas is somewhat problematic.  Some of them can be eliminated without trouble, but some of the warnings are nonlocal: they appear not in the time stepper but in the class used to represent the data.  For now that means some random function in the test, but this presumably means we will get warnings from DataVector or something when we try to use this in a simulation.

GCC bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80635

**Comment by kidder** [comment]
So, to be sure I understand, you are talking about the maybe-uninitialized warnings caused by the use of boost::optional.  Or are there other warnings?  Given the link you posted, I would suggest turning the maybe-uninitialized warnings off via compiler flag rather than pragmas if this is problematic

**Comment by wthrowe** [comment]
That was what I was talking about, but I've realized I can avoid the optional by using make_arithmetic_value or the thing it's being replaced with, so silencing the warnings shouldn't be necessary.

**Comment by wthrowe** [comment]
The fact that there are unsuppressable clang-tidy warnings that you can't fix is apparently intentional and not considered a bug: https://reviews.llvm.org/D26418#655058 (bullet 3).

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 29]
```diff
@@ -354,75 +481,55 @@ BoundaryVars AdamsBashforthN::compute_boundary_delta(
...
-                                                simulation_less));
+      side_indices.push_back(step_containing_time(
+          side_hist, *union_times_for_coupling_begin, simulation_less));
     }
 
-    for (auto smallest_scale_time = smallest_scale_times_begin;
-         smallest_scale_time != smallest_scale_times_end;
-         ++smallest_scale_time) {
+    // This loop computes the coefficient of the coupling evaluation
+    // as deriv_coef.
```
Comment (kidder) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: (i.e. the sum of the capital i s)

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 30]
```diff
@@ -70,6 +70,98 @@ class AdamsBashforthN : public TimeStepper::Inherit {
       const std::deque<std::tuple<Time, Vars, DerivVars>>& history,
       const TimeDelta& time_step) const noexcept;
 
+  /*!
+   * An explanation of the computation being performed by this
```
Comment (kidder) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp:74]: define k as the order of AB

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**